### PR TITLE
Private-Events

### PR DESCRIPTION
**Pull request by warblo001**
# Private-Events

This project is about building an exclusive events site where users can sign up with a username, create and attend events.

## Events project features:

- Users can make and attend events.

## Built With

- Ruby on Rails
- Bulma

### CODE REVIEW

File: app/views/events/index.html.erb [hunk 1]
```diff
@@ -1,33 +1,51 @@
...
+    <div class="columns">
+      <% @events.upcoming.each do |event| %>
+        <div class="column is-3">
+          <div class="card">
+            <header class="card-header">
+              <%= link_to event.name, event_path(event), class: "card-header-title is-centered" %>
+            </header>
+            <div class="card-content">
+              <div class="content has-text-justified">
+                <%= event.description.truncate_words(18) %>
```
Comment (rukundoeric) [inline app/views/events/index.html.erb]: this logic shouldn't be put in the views. Move it to the controller or helper.

File: app/views/events/index.html.erb [hunk 2]
```diff
@@ -1,33 +1,51 @@
...
+    <div class="columns">
+      <% @events.past.each do |event| %>
+        <div class="column is-3">
+          <div class="card">
+            <header class="card-header">
+              <%= link_to event.name, event_path(event), class: "card-header-title is-centered" %>
+            </header>
+            <div class="card-content">
+              <div class="content has-text-justified">
+                <%= event.description.truncate_words(18) %>
```
Comment (rukundoeric) [inline app/views/events/index.html.erb]: this logic shouldn't be put in the views. Move it to the controller or helper.

File: app/views/events/show.html.erb [hunk 3]
```diff
@@ -1,24 +1,97 @@
...
+                <div class="dropdown-menu" id="dropdown-menu6" role="menu">
+                  <div class="dropdown-content">
+                    <ul>
+                      <% if User.find(@event.attendances.map { |rol| rol.user_id }).any? %>
+                        <div class="dropdown-item">
+                          <% User.find(@event.attendances.map { |rol| rol.user_id }).each do |user| %>
+                            <li><p><%= user.name %></p></li>
+                          <% end %>
+                        </div>
+                      <% end %>
```
Comment (rukundoeric) [inline app/views/events/show.html.erb]: this logic shouldn't be put in the views. Move it to the controller or helper.

File: app/views/events/show.html.erb [hunk 4]
```diff
@@ -1,24 +1,97 @@
...
+                  </div>
+                </div>
+              </div>
+            </div>
+
+            <div class="column">
+              <div class="dropdown is-hoverable">
+                <div class="dropdown-trigger">
+                  <button class="button is-fullwidth" aria-haspopup="true" aria-controls="dropdown-menu3">
+                    <span>Attendies: <%= User.find(@event.attendances.map { |rol| rol.user_id }).count %></span>
```
Comment (rukundoeric) [inline app/views/events/show.html.erb]: this logic shouldn't be put in the views. Move it to the controller or helper.

File: app/views/events/show.html.erb [hunk 5]
```diff
@@ -1,24 +1,97 @@
...
+                <div class="dropdown-menu" id="dropdown-menu6" role="menu">
+                  <div class="dropdown-content">
+                    <ul>
+                      <% if User.find(@event.attendances.map { |rol| rol.user_id }).any? %>
+                        <div class="dropdown-item">
+                          <% User.find(@event.attendances.map { |rol| rol.user_id }).each do |user| %>
+                            <li><p><%= user.username %></p></li>
+                          <% end %>
+                        </div>
+                      <% end %>
```
Comment (rukundoeric) [inline app/views/events/show.html.erb]: this logic shouldn't be put in the views. Move it to the controller or helper.

File: app/views/events/show.html.erb [hunk 6]
```diff
@@ -1,24 +1,97 @@
...
+          <% if @event.user == current_user %>
+            <%= button_to "Invite", users_path, params: {:event_id => @event.id}, method: :get, class: "button is-info is-large is-fullwidth" %>
+          <% else %>
+            <% attendance = Attendance.find_by(event_id: @event.id, user_id: current_user.id) %>
+            <% if attendance && attendance.accepted? %>
+              <%= button_to "Cancel attendance", attendance_path(id: attendance.id, event_id: @event.id, user_id: current_user.id), method: :delete, class: "button is-danger is-large is-fullwidth" %>
+            <% else %>
+              <%= button_to "Attend", attendance_path(event_id: @event.id), method: :patch, class: "button is-primary is-large is-fullwidth" %>
+            <% end %>
+          <% end %>
```
Comment (rukundoeric) [inline app/views/events/show.html.erb]: this logic shouldn't be put in the views. Move it to the controller or helper.

File: app/views/users/index.html.erb [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+          <h2 class="title is-2 is-spaced has-text-centered mb-6">Guest List</h2>
+    
+          <div class="columns is-centered">
+            <table class="table is-fullwidth is-hoverable">
+              <tbody>
+                <th class="has-text-centered">Name</th>
+                <th class="has-text-centered">Status</th>
+                <th class="has-text-centered">Action</th>
+                <% @users.reject { |user| user == Event.find(params[:event_id]).creator }.each do |user| %>
+                  <% attendance = Attendance.find_by(event_id: params[:event_id], user_id: user.id) %>
```
Comment (rukundoeric) [inline app/views/users/index.html.erb]: this logic shouldn't be put in the views. Move it to the controller or helper.

File: app/views/users/show.html.erb [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+	      <div class="columns is-centered">
+          <div class="column has-text-centered">
+            <h3 class="subtitle is-3 has-text-centered">Hosted Events: <%= @user.events.count %></h3>
+            <ul>
+              <% if @user.events.empty? %>
+                <p>User is not hosting</p>
+              <% else %>
+                <% @user.events.each do |event| %>
+                  <li class="has-text-centered"><%=link_to event.name, event_path(event) %></li>
+                <% end %>
```
Comment (rukundoeric) [inline app/views/users/show.html.erb]: this logic shouldn't be put in the views. Move it to the controller or helper.

File: app/views/users/show.html.erb [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+              <% end %>
+            </ul>
+          </div>
+
+          <div class="column has-text-centered">
+            <h3 class="subtitle is-3">Upcoming Events: <%= @user.attended_events.where('date >= ?', Date.today).count %></h3>
```
Comment (rukundoeric) [inline app/views/users/show.html.erb]: this logic shouldn't be put in the views. Move it to the controller or helper.

File: app/views/users/show.html.erb [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+              <% else %>
+                <% @user.attended_events.where('date >= ?', Date.today).each do |event| %>
+                  <li class="has-text-centered"><%=link_to event.name, event_path(event) %></li>
+                <% end %>
+              <% end %>
+            </ul>
+          </div>
+
+          <div class="column has-text-centered">
+            <h3 class="subtitle is-3 has-text-centered">Attended Events: <%= @user.attended_events.where('date < ?', Date.today).count %></h3>
```
Comment (rukundoeric) [inline app/views/users/show.html.erb]: this logic shouldn't be put in the views. Move it to the controller or helper.

File: app/views/users/show.html.erb [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+            <ul>
+              <% if @user.attended_events.empty? %>
+                <p>User is not partying</p>
+              <% else %>
+                <% @user.attended_events.where('date < ?', Date.today).each do |event| %>
+                  <li class="has-text-centered"><%=link_to event.name, event_path(event) %></li>
+                <% end %>
```
Comment (rukundoeric) [inline app/views/users/show.html.erb]: this logic shouldn't be put in the views. Move it to the controller or helper.

**Review by rukundoeric — CHANGES_REQUESTED** [review]
# Status: ♻️ CHANGES REQUIRED ♻️
_Great job guys so far 👏 💯 👍. However, there are few issues we need to address before we move on, please check the comments below._

- ### To Highlight
  - [x] The project is clean. 💯
  - [x] The project follows Git flow. 💯

- ### To Improve

  - [x] [MANDATORY] Only logged-in users should be able to see create-event page. In your case, any user can access it.
  - [x] [MANDATORY] After creating Event, I clicked on Invite button and I got this error: ```undefined method `creator' for #<Event:0x00007fffce9ee708>
Did you mean?  create_user```.
  - [x] [MANDATORY] When I open the event created by someone else and I click on attend button, I get this error: ```uninitialized constant AttendancesController Did you mean? AttendanceController```.
  - [x] [MANDATORY] After creating an event, User should be one of the attendees automatically.
  - [x] [MANDATORY] I could not find any test, kindly implement them.
  - [x] [MANDATORY] All logic related code that will be used in views, should be put either in controllers or helpers. 

_Please make all the required changes before you submit another code review. If you have any problems or concerns, you can reach out to me on Slack: Rukundo Eric 🚀 . I wish you all the very best 🥇_

### _Keep working hard._
### _Happy Coding😃_
-----------------------------------------------------------------------------------------------------------------------------------------
**Regards,
[Rukundo Eric](https://rukundo-eric.herokuapp.com/) ,
[LinkedIn](https://www.linkedin.com/in/rukundo-eric/), 
Slack: Rukundo Eric**

**Review by cyonii — CHANGES_REQUESTED** [review]
# Status: Changes Required ♻️✍️

Hello @warblo001 , you have shown a great effort so far, keep up the good energy. 

## To Improve
- [x] As the previous reviewer pointed out, only authenticated users should be able to see the event-create page, your app currently allows unauthenticated users to access it. Please endeavor to implement this change.
- [x] Again just like the previous Reviewer stated when I click the `Invite` button after creating an event, your app throws this error
```ruby
undefined method `creator' for #<Event>
Did you mean?  create_user
```
Your `Event` model doesn't appear to have a `creator` association. Please revisit the specs and instructions for this project, [here](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49723079/)
- [x] When I view an event created by someone else and click "Attend", it throws this error saying
```
uninitialized constant AttendancesController Did you mean? AttendanceController
```
- [x] You have not addressed the inline comments where the previous TSE asked you to move your application logic out of your views, you should use helpers instead.


## Nice to have (Optional)
Note that these are optional suggestions, while they are not mandatory, I recommend that you implement them to improve your project.
- [x] I am unable to log in back after O log out, it says `Name is invalid`, please fix this so that users can sign in and out
- [ ] Update your `Gemfile.lock` by running `bundle install`, `mimemagic` gem is still locked while it no longer appears in your `Gemfile`

> It appears you have not addressed any of the issues pointed out by the previous Reviewer, I recommend you fix these issues based on the first review. If something is blocking you, or this is some kind of mistake, please feel free to respond here.

Happy coding, see you at the top

> #### <br>You can ping me on Slack @[CY Kalu](https://microverse-students.slack.com/team/U01494U96FR) to draw my attention in case I do not respond quickly on here. I'm always available and will be glad to help.<br><br>

**Review by smcommits — CHANGES_REQUESTED** [review]
# Status: Changes Requested 
## Hello! You guys have done a great job. The application appears to be fantastic and the project is almost complete. To make it even better, we still need to address some issues that were pointed out by the previous TSE. 

## Changes Required

#### You still haven't addressed the following issues.

- [x] **Only authenticated users should be able to visit the page to create a new event**

User can visit the page to create an event without any authentication
:-----------------------------------------------------------------------------------------:
create


I can see that you are almost there, great work! You have the helper methods in the application controller to require login. You also need to use before_action to append a callback before respective actions in your events controller. Please use [this link](https://api.rubyonrails.org/v6.1.3.1/classes/AbstractController/Callbacks/ClassMethods.html#method-i-before_action) to understand better. 

<hr>

- [x] **Invite button is still throwing an error. Please fix it. Refer to the image below to understand more**

Error upon clicking the invite button
:---------------------------------------------:
invite

Note: Rails follows a naming convention. Please use [this gist](https://gist.github.com/iangreenleaf/b206d09c587e8fc6399e) to understand more. 

<hr>

- [x] **You still have logic in your views. Please refer to the inline comments by @rukundoeric**

This is very important. You should not have a lot of logic in your views. The view is a presentational layer and having a lot of logic is a bad idea for various reasons. For instance, you can soon end up with a large file that is a mix of HTML and Ruby code that can be difficult to manage and maintain. This is also an area that can lead to lots of repetition, leading to violations of DRY (don’t repeat yourself) principles.

# 
### Please don't hesitate to contact me on [Slack](https://microverse-students.slack.com/team/U017N7LLZ40) if you need assistance or believe I have overlooked something and made an unfair request.

**Review by descholar-ceo — CHANGES_REQUESTED** [review]
# Status: Changes Required :vertical_traffic_light: :warning: 
## Hello @warblo001  :handshake: :handshake: :handshake: 

:+1: This is good to make it this far, your project is really nice, congratulations on the good work you have done so far.
:+1: I liked that you have tried to implement the requested changes from the previous review, keep it up, and your project is really almost done.
#### :warning: However we need to make few changes to mark this project approved, don't feel discouraged because of being asked some changes more and more, that's beacuse this is a learning process, and we are here to help you to become an out-standing developer. See the requested change below :point_down: :point_down: :point_down: 
### Note: All of requested changes which are marked as [Required] are mandatory to be implemented in order to get this project approved, others which are marked as [Optional] are nice to have in your project to make it outstanding

# [Required Changes] :red_circle: 
- [x] There are still logics in your views, kindly read and address to the inline comments of @rukundoeric as he showed you wherever there are logics
- [x] After clicking on invite to event button and I invite an attendee to an event, I can't go back to the show page of that event, that's a critical issue which is making your app to crash, please fix it.
Screenshot from 2021-04-04 13-14-13

___
_If anything seems to be unclear to you, reachout and ask anything by commenting on this Pull Request and mention me as [descholar-ceo](https://github.com/descholar-ceo), I will receive a notification and will respond._

[GitHub](https://github.com/descholar-ceo) | [LinkedIn](https://linkedin.com/in/mugirase-emmanuel) | [Twitter](https://twitter.com/descholar3)

_Reviewed by [Mugirase Emmanuel (descholar)](https://github.com/descholar-ceo)_

**Review by rindrajosia — CHANGES_REQUESTED** [review]
Changes Required :red_circle:
To Highlight: :angel:

- Good pull request. :+1: 
- Good readme file :100: 

To Improve :factory_worker: (REQUIRED)

- [ ] EventsController#index action:

1. Lists past events, i.e. @past_events = Event.previous_events ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/past_upcoming_events_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))
2. lists ucomping events, i.e. @upcoming_events = Event.upcoming_events ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/past_upcoming_events_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))

- [ ] UsersController#show action:

1. lists events created by user, i.e. @created_events = current_user.events ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/current_user_show_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))
2.  lists events previoulsy attendented by user **using event model scope**, i.e. @past_events = current_user.attended_events.previous ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/user_events_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))
3. lists upcoming events previously attended by user **using event model scope**, i.e. @upcoming_events = current_user.attended_events.upcoming ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/user_events_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))

- [ ] It seems that your app doesn't allow me to sign in with other users account:


- [ ] User has difficulty to attend someone else events:


- [ ] I got error when I click on past events created by someone else:


- [ ] Please, make sure that all of your test units don't fail, it will be nice if you add more test units:


- [ ] It will be perfect if you remove also logic in your view and partial files and put it in your helpers. As a developer your job is to create quality code that solves your client’s problems. Quality has many aspects to it, including readability and maintainability. It is much easier to read code that is lean and skinny, as well as serving a single purpose. When you look at a CSS file, you know it’s for styling and giving your site personality and uniqueness. When you look at a JavaScript file, you know it handles the interactivity on the front-end for the person viewing the site. And the view file is responsible for the HTML.

That's it, you can submit another code review request after making these changes :recycle: :innocent:
With all my respect :pray:
Rindra Josia

File: app/views/layouts/application.html.erb [hunk 12]
```diff
@@ -11,6 +11,60 @@
...
+  <% if flash[:notice] %>
+      <div class="notification is-primary global-notification">
+        <p class="notice"><%= notice %></p>
+      </div>
+    <% end %>
+    <% if flash[:alert] %>
+      <div class="notification is-danger global-notification">
+        <p class="alert"><%= alert %></p>
+      </div>
+    <% end %>
```
Comment (oloomoses) [inline app/views/layouts/application.html.erb]: - [ ] Kindly Move this logic into helpers

File: app/views/layouts/application.html.erb [hunk 13]
```diff
@@ -11,6 +11,60 @@
...
+                  <%= link_to "Logout", logout_path, class: "button is-white" %>
+                </p>
+              <% else %>
+                <p class="control">
+                  <%= link_to 'Sign In', login_path, class: "button is-white" %>
+                </p>
+						    <p class="control">
+                  <%= link_to 'Sign Up', signup_path, class: "button is-white" %>
+                </p>
+              <% end %>
```
Comment (oloomoses) [inline app/views/layouts/application.html.erb]: - [ ] Kindly Move this logic into helpers

File: app/views/users/index.html.erb [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+                    <% end %>
+                    <% if attendance.nil? %>
+                      <td class="has-text-centered">
+                        <%= button_to "Invite", attendances_path(event_id: params[:event_id], user_id: user.id), class: "button is-primary is-small" %>
+                      </td>
+                    <% else %>
+                      <td class="has-text-centered">
+                        <%= button_to "Cancel Invitation", attendance_path(id: attendance.id, event_id: params[:event_id]), method: :delete, class: "button is-danger is-small" %>
+                      </td>
+                    <% end %>
```
Comment (oloomoses) [inline app/views/users/index.html.erb:35]: - [ ] Kindly Move this logic into helpers

File: app/views/users/show.html.erb [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+	      <div class="columns is-centered">
+          <div class="column has-text-centered">
+            <h3 class="subtitle is-3 has-text-centered">Hosted Events: <%= @user.events.count %></h3>
+            <ul>
+              <% if @user.events.empty? %>
+                <p>User is not hosting</p>
+              <% else %>
+                <% @user.events.each do |event| %>
+                  <li class="has-text-centered"><%= link_toevent %></li>
+                <% end %>
```
Comment (oloomoses) [inline app/views/users/show.html.erb:15]: - [ ] Kindly Move this logic into helpers

File: app/views/users/show.html.erb [hunk 16]
```diff
@@ -0,0 +1,49 @@
...
+
+          <div class="column has-text-centered">
+            <h3 class="subtitle is-3">Upcoming Events: <%= upcomingev.count %></h3>
+            <ul>
+              <% if upcomingev.count.zero? %>
+                <p>No planned parties</p>
+              <% else %>
+                <% upcomingev.each do |event| %>
+                  <li class="has-text-centered"><%= link_toevent %></li>
+                <% end %>
```
Comment (oloomoses) [inline app/views/users/show.html.erb:28]: - [ ] Kindly Move this logic into helpers

File: app/views/users/show.html.erb [hunk 17]
```diff
@@ -0,0 +1,49 @@
...
+
+          <div class="column has-text-centered">
+            <h3 class="subtitle is-3 has-text-centered">Attended Events: <%= attevents.count %></h3>
+            <ul>
+              <% if @user.attended_events.empty? %>
+                <p>User is not partying</p>
+              <% else %>
+                <% attevents.each do |event| %>
+                  <li class="has-text-centered"><%= link_toevent %></li>
+                <% end %>
```
Comment (oloomoses) [inline app/views/users/show.html.erb:41]: - [ ] Kindly Move this logic into helpers

File: app/views/events/show.html.erb [hunk 18]
```diff
@@ -1,24 +1,102 @@
...
+                    </button>
+                  </div>
+                <div class="dropdown-menu" id="dropdown-menu6" role="menu">
+                  <div class="dropdown-content">
+                    <ul>
+                      <% if userfinmap.any? %>
+                        <div class="dropdown-item">
+                          <% userfinmap.each do |user| %>
+                            <li><p><%= user.username %></p></li>
+                          <% end %>
```
Comment (oloomoses) [inline app/views/events/show.html.erb:28]: - [ ] Kindly Move this logic into helpers

File: app/views/events/show.html.erb [hunk 19]
```diff
@@ -1,24 +1,102 @@
...
+                  </button>
+                </div>
+                <div class="dropdown-menu" id="dropdown-menu6" role="menu">
+                  <div class="dropdown-content">
+                    <ul>
+                      <% if userfinmap.any? %>
+                        <div class="dropdown-item">
+                          <% userfinmap.each do |user| %>
+                            <li><p><%= user.username %></p></li>
+                          <% end %>
```
Comment (oloomoses) [inline app/views/events/show.html.erb:51]: - [ ] Kindly Move this logic into helpers

File: app/views/events/show.html.erb [hunk 20]
```diff
@@ -1,24 +1,102 @@
...
+          <% if @event.user == current_user %>
+            <%= button_to "Invite", users_path, params: {:event_id => @event.id}, method: :get, class: "button is-info is-large is-fullwidth" %>
+          <% else %>
+            <% attfindb %>
+            <% if attfindb  %>
+              <%= button_to "Cancel attendance", attendance_path(id: attendance.id, event_id: @event.id, user_id: current_user.id), method: :delete, class: "button is-danger is-large is-fullwidth" %>
+            <% else %>
+              <%= link_toattend %>
+            <% end %>
+          <% end %>
```
Comment (oloomoses) [inline app/views/events/show.html.erb:91]: - [ ] Kindly Move this logic into helpers

**Review by oloomoses — CHANGES_REQUESTED** [review]
# Changes Required :recycle: 
Hello team, you have done great implementing some of the changes from the previous reviews. :muscle: 
However, there are a few mandatory changes that you have to make for this project to be approved.

- [ ] I Just signed up for the app without filling the form. This is wrong. Kindly add proper validations to your models and ensure that a user cannot signup without entering the username in the form.
- [ ] Some of your tests are still failing. Kindly fix them.
- [ ] You still have lots of logic in your views, check the comments. Kindly move them into helpers. [Here is a guide to help through this](https://medium.com/@acushla4real/rails-view-helpers-best-practices-guide-d7d7b6db3db4) in case you feel stuck.

**Kindly implement the above changes before submitting another review.**
**Feel free to reach me on slack or leave a comment while mentioning me and I will be happy to reply**

_Slack: Oloo Moses_

**Comment by mimipeshy** [comment]
## STATUS: CHANGES REQUIRED 🔴 ♻️ 
Hello there @warblo001 I appreciate the effort you have put into this project. It wonderful you have taken the time to work on the comments given by the TSE's. However,there is still room for improvement and one major thing is still missing:

- [x] You have a failing test as shown in the image below kindly fix it. 

image

**[OPTIONAL]**

- [ ] As the previous TSE noted, you still have lots of logic in your views, check his comments. [Here is a guide to help through this](https://medium.com/@acushla4real/rails-view-helpers-best-practices-guide-d7d7b6db3db4) in case you feel stuck. Kindly take time and go through this article and understand the importance of helpers. You will use them a lot in this module. Feel free to reach out to any TSE if you are finding it hard to understand the topic, we will be glad to help

That's all :smile: kindly submit another review when done.

_Feel free to reach out to me on slack if you have any questions or something is unclear, or just comment on this PR and tag my name  (@mimipeshy) and I will be glad to help_
Regards
Peris Ndanu

**Review by smcommits — APPROVED** [review]
# Status: Approved :+1: 
## Hello, team! The application looks fantastic. Great job. Good work addressing the issues that previous TSEs pointed out. This PR is now approved. You can go ahead and merge the branch. Congratulations and good luck with your next project.

**Pull request merged**